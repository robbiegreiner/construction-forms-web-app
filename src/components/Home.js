import React, { Component } from 'react';
import { Route } from 'react-router-dom';
import '../styles/Home.css';
import DataViz from './DataViz.js';
import Controls from './Controls.js';

class Home extends Component {
  
  constructor() {
    super();

    this.state = {
      selectedControl: '',
      newEmployee: {}
    };
  }

  selectControl = (control) => {
    this.setState({selectedControl: control});
  }

  handleEmployeeChange = (event) => {
    const newK = event.target.className
    const newO = Object.assign({}, this.state.newEmployee, { [event.target.className]: event.target.value })
    this.setState({
      newEmployee: newO
    })
  }

  postNewEmployee = (event) => {
    event.preventDefault();
    const postBody = Object.assign({},
      {
        name: this.state.newEmployee.employee_name,
        position: this.state.newEmployee.employee_position,
        email: this.state.newEmployee.employee_email,
        phone: this.state.newEmployee.employee_phone,
      }
    );
    fetch(`http://localhost:4000/api/v1/employees`, {
      method: 'POST',
      headers: {
        'Content-type': 'application/json'
      },
      body: JSON.stringify(postBody)
    })
      .then(res => res.json())
      .catch(error => { throw error; });
  }


  renderPopUp() {
    if (this.state.selectedControl === 'Add Employee') {
      return (
        <div className="add-popup">
          <form className="popup-form" onSubmit={this.postNewEmployee}>
            <button 
              onClick={() => this.setState({ selectedControl: '' })} 
              className="close-popup"
            >X</button>
            <h4>Add New Employee</h4>
            <label htmlFor="employee-name">Name: 
              <input 
                type="text" 
                ref="employee-name" 
                className="employee_name"  
                placeholder="Name" 
                onChange={this.handleEmployeeChange}
              />
            </label>
            <label htmlFor="employee-position">Position: 
              <input 
                type="text" 
                ref="employee-position" 
                className="employee_position"  
                placeholder="Position" 
                onChange={this.handleEmployeeChange}
              />
            </label>
            <label htmlFor="employee-email">Email: 
              <input 
                type="email" 
                ref="employee-email" 
                className="employee_email"  
                placeholder="Email" 
                onChange={this.handleEmployeeChange}
              />
            </label>
            <label htmlFor="employee-phone">Phone: 
              <input 
                type="text" 
                ref="employee-phone" 
                className="employee_phone"  
                placeholder="Phone" 
                onChange={this.handleEmployeeChange}
              />
            </label>
            <input type="submit" value="Submit" />
          </form>
        </div>
      )
    }
  }

  postNewEmployee(event) {
    event.preventDefault();
    console.log(event.target) 
  }

  render() {
    return (
        <div className="main-container">
          <Controls 
            selectControl = {this.selectControl}
          />
          <DataViz 
            selectedControl = {this.state.selectedControl}
          />
          {this.renderPopUp()}
        </div>
    );
  }
}

export default Home;
