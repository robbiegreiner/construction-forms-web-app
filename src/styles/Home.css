.main-container {
  width: 90vw;
  height: 80vh;
  background-color: white;
  border-radius: 4px;
  display: flex;
  position: relative;
  box-shadow: 0px 8px 24px 0px rgba(0,0,0,0.19);
}

.add-popup {
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0,0,0,.5);
}

.add-popup label {
  display: block;
}

.add-popup input {
  display: block;
  width: 80%;
  margin: auto;
}

.popup-form {
  text-align: center;
  padding: 25px;
  width: 50%;
  background-color: #DEE3EB;
}

.close-popup {
  float: right;
}
