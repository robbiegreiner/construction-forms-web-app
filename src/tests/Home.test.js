/* eslint-disable no-unused-vars, no-undef */
import React from 'react';
import { shallow, mount } from 'enzyme';
import Home from '../components/Home';

describe('Home tests', () => {

  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<Home />);
  });

  it('should render without crashing', () => {
    shallow(<Home />);
  });

  it('should render with three children', () => {
    expect(wrapper.find('div').getElements()[0].props.children.length).toEqual(3);
  });

  it('should render with empty state', () => {
    const emptyState = {
      selectedControl: '',
      newEmployee: {
        employee_name: '',
        employee_position: '',
        employee_phone: '',
        employee_email: ''
      },
      newProject: { 
        project_name: '',
        project_location: '',
        project_public: false,
        project_union: false
      }
    };

    expect(wrapper.state()).toEqual(emptyState);
  });

  it('should be able to select a control', () => {
    wrapper.instance().selectControl('Employees');
    expect(wrapper.state().selectedControl).toEqual('Employees');
  });

  it('should be able to deselect a control if the same input is entered twice', () => {
    wrapper.instance().selectControl('Employees');
    wrapper.instance().selectControl('Employees');
    expect(wrapper.state().selectedControl).toEqual('');
  });

  it('should handle change on the new employee input form', () => {
    
  });

  it('should match the snapshot test', () => {
    expect(wrapper).toMatchSnapshot();
  });

});
