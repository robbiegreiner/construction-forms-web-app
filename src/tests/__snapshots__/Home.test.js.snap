// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home tests should match the snapshot test 1`] = `
<div
  className="main-container"
>
  <Controls
    selectControl={[Function]}
    selectedControl=""
  />
  <DataViz
    selectedControl=""
  />
</div>
`;

exports[`Home tests should render a form to add a new employee 1`] = `
<div
  className="main-container"
>
  <Controls
    selectControl={[Function]}
    selectedControl="Add Employee"
  />
  <DataViz
    selectedControl="Add Employee"
  />
  <div
    className="add-popup"
  >
    <form
      className="popup-form"
      onSubmit={[Function]}
    >
      <button
        className="close-popup"
        onClick={[Function]}
      >
        X
      </button>
      <h4>
        Add New Employee
      </h4>
      <label
        htmlFor="employee-name"
      >
        Name:
        <input
          className="employee_name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
        />
      </label>
      <label
        htmlFor="employee-position"
      >
        Position:
        <input
          className="employee_position"
          onChange={[Function]}
          placeholder="Position"
          type="text"
        />
      </label>
      <label
        htmlFor="employee-email"
      >
        Email:
        <input
          className="employee_email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
        />
      </label>
      <label
        htmlFor="employee-phone"
      >
        Phone:
        <input
          className="employee_phone"
          onChange={[Function]}
          placeholder="Phone"
          type="text"
        />
      </label>
      <input
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</div>
`;

exports[`Home tests should render a form to add a new project 1`] = `
<div
  className="main-container"
>
  <Controls
    selectControl={[Function]}
    selectedControl="Add Project"
  />
  <DataViz
    selectedControl="Add Project"
  />
  <div
    className="add-popup"
  >
    <form
      className="popup-form"
      onSubmit={[Function]}
    >
      <button
        className="close-popup"
        onClick={[Function]}
      >
        X
      </button>
      <h4>
        Add New Project
      </h4>
      <label
        htmlFor="project-name"
      >
        Project Name:
        <input
          className="project_name"
          onChange={[Function]}
          placeholder="Project Name"
          type="text"
        />
      </label>
      <label
        htmlFor="project-location"
      >
        Location:
        <input
          className="project_location"
          onChange={[Function]}
          placeholder="Location"
          type="text"
        />
      </label>
      <label
        htmlFor="project-union"
      >
        Union:
        <select
          className="project_union"
          onChange={[Function]}
        >
          <option
            value="true"
          >
            Yes
          </option>
          <option
            value="false"
          >
            No
          </option>
        </select>
      </label>
      <label
        htmlFor="project-public"
      >
        Public:
        <select
          className="project_public"
          onChange={[Function]}
        >
          <option
            value="true"
          >
            Yes
          </option>
          <option
            value="false"
          >
            No
          </option>
        </select>
      </label>
      <input
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</div>
`;
